contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -1,26 +1,148 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
+import { StableMath } from "../utils/StableMath.sol";
 import { VaultCore } from "./VaultCore.sol";
 
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IStrategy } from "../interfaces/IStrategy.sol";
+
 /**
  * @title OETH VaultCore Contract
  * @author Origin Protocol Inc
  */
 contract OETHVaultCore is VaultCore {
+    using SafeERC20 for IERC20;
+    using StableMath for uint256;
+
     address public immutable weth;
+    uint256 public wethAssetIndex;
 
     constructor(address _weth) {
         weth = _weth;
     }
 
+    /**
+     * @dev Caches WETH's index in `allAssets` variable.
+     *      Reduces gas usage by redeem by caching that.
+     */
+    function cacheWETHAssetIndex() external onlyGovernor {
+        uint256 assetCount = allAssets.length;
+        for (uint256 i = 0; i < assetCount; ++i) {
+            if (allAssets[i] == weth) {
+                wethAssetIndex = i;
+                break;
+            }
+        }
+
+        require(allAssets[wethAssetIndex] == weth, "Invalid WETH Asset Index");
+    }
+
     // @inheritdoc VaultCore
-    function mint(
+    function _mint(
         address _asset,
         uint256 _amount,
         uint256 _minimumOusdAmount
-    ) external virtual override whenNotCapitalPaused nonReentrant {
+    ) internal virtual override {
         require(_asset == weth, "Unsupported asset for minting");
-        _mint(_asset, _amount, _minimumOusdAmount);
+        require(_amount > 0, "Amount must be greater than 0");
+        require(
+            _amount >= _minimumOusdAmount,
+            "Mint amount lower than minimum"
+        );
+
+        emit Mint(msg.sender, _amount);
+
+        // Rebase must happen before any transfers occur.
+        if (!rebasePaused && _amount >= rebaseThreshold) {
+            _rebase();
+        }
+
+        // Mint oTokens
+        oUSD.mint(msg.sender, _amount);
+
+        // Transfer the deposited coins to the vault
+        IERC20(_asset).safeTransferFrom(msg.sender, address(this), _amount);
+
+        // Auto-allocate if necessary
+        if (_amount >= autoAllocateThreshold) {
+            _allocate();
+        }
+    }
+
+    // @inheritdoc VaultCore
+    function _calculateRedeemOutputs(uint256 _amount)
+        internal
+        view
+        virtual
```

Comment:
nit `virtual` is not needed here as `OETHVaultCore` is not inherited. Only the `override` is required